In [1]:
from IPython.core.display import display, HTML

# 노트북 컨테이너 크기 조절
display(HTML("<style>#notebook-container {width: 90% !important; }</style>"))
# 마크다운 폰트
display(HTML("<style>.rendered_html{font-family: 'Noto Sans KR'; font-size: 18pt; line-height: 140%;}</style>"))
# DataFrame 테이블 폰트
display(HTML("<style>.rendered_html table{font-family: 'Noto Sans KR'; font-size: 18pt; line-height: 140%;}</style>"))
# 출력창 폰트
display(HTML("<style>pre{font-family: 'Noto Sans KR', font-size: 18pt;}.rendered_html>ul>li{margin: 10px}</style>"))
display(HTML("<style>div.output_area>pre, div.output_subarea>pre{font-size: 18pt !important; line-height: 140%;}</style>"))
# 코드셀 폰트
display(HTML("<style> .CodeMirror pre {font-family: 'Verdana'; font-size: 20pt !important; line-height: 140%;}</style>"))

In [2]:
import pandas as pd
import numpy as np
import datetime
import seaborn as sns
import warnings

warnings.filterwarnings('ignore')

In [4]:
# 타이타닉 데이터셋 로드
df = sns.load_dataset('titanic')
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [41]:
deck = df['deck'].cat.codes
deck.loc[deck >= 0].mean()

2.3694581280788176

In [5]:
# 결측치 확인
df.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

## age의 결측치를 채웁니다.

In [7]:
# 코드를 입력해 주세요
df['age'] = df['age'].fillna(df['age'].mean())

In [8]:
# 검증코드
df['age'].isnull().sum()

0

<p><strong>[출력 결과]</strong></p><pre>0</pre>

## Embarked 결측치를 채웁니다.

In [12]:
# 코드를 입력해 주세요
df['embarked'] = df['embarked'].fillna('S')

In [13]:
# 검증코드
df['embarked'].isnull().sum()

0

<p><strong>[출력 결과]</strong></p><pre>0</pre>

## sex 컬럼을 encoding 변환하여 남자는 1 여자는 0으로 변환합니다.

In [17]:
# 코드를 입력해 주세요
df['sex'] = df['sex'].apply(lambda x: 1 if x == 'male' else 0)

In [18]:
# 검증코드
df['sex']

0      1
1      0
2      0
3      0
4      1
      ..
886    1
887    0
888    0
889    1
890    1
Name: sex, Length: 891, dtype: int64

<p><strong>[출력 결과]</strong></p><pre>0      1
1      0
2      0
3      0
4      1
      ..
886    1
887    0
888    0
889    1
890    1
Name: sex, Length: 891, dtype: int64</pre>

## embarked 컬럼을 encoding 변환합니다.

apply 를 사용합니다.
- 'S': 0
- 'C': 1
- 'Q': 2

In [21]:
m = {
    'S': 0, 
    'C': 1, 
    'Q': 2
}
m

{'S': 0, 'C': 1, 'Q': 2}

In [34]:
def convert(x):
    if x == 'S':
        return 0
    elif x == 'C':
        return 1
    elif x == 'Q':
        return 2

In [35]:
df['embarked'].apply(convert)

0      None
1      None
2      None
3      None
4      None
       ... 
886    None
887    None
888    None
889    None
890    None
Name: embarked, Length: 891, dtype: object

In [32]:
df['embarked']

0      0
1      1
2      0
3      0
4      0
      ..
886    0
887    0
888    0
889    1
890    2
Name: embarked, Length: 891, dtype: int64

In [23]:
# 코드를 입력해 주세요
df['embarked'] = df['embarked'].map(m)

In [24]:
# 검증코드
df['embarked'].value_counts()

0    646
1    168
2     77
Name: embarked, dtype: int64

<p><strong>[출력 결과]</strong></p><pre>0    646
1    168
2     77
Name: embarked, dtype: int64</pre>

In [25]:
df.head()

survived  pclass  sex   age  sibsp  parch     fare  embarked  class    who  \
0         0       3    1  22.0      1      0   7.2500         0  Third    man   
1         1       1    0  38.0      1      0  71.2833         1  First  woman   
2         1       3    0  26.0      0      0   7.9250         0  Third  woman   
3         1       1    0  35.0      1      0  53.1000         0  First  woman   
4         0       3    1  35.0      0      0   8.0500         0  Third    man   

   adult_male deck  embark_town alive  alone  
0        True  NaN  Southampton    no  False  
1       False    C    Cherbourg   yes  False  
2       False  NaN  Southampton   yes   True  
3       False    C  Southampton   yes  False  
4        True  NaN  Southampton    no   True

## Who

In [51]:
df['who'].value_counts()

man      537
woman    271
child     83
Name: who, dtype: int64

In [52]:
who_m = {
    'man': 0, 
    'woman': 1, 
    'child': 2
}

In [54]:
df['who'] = df['who'].map(who_m)

In [55]:
df['who'].value_counts()

0    537
1    271
2     83
Name: who, dtype: int64

## features, labels 정의

In [56]:
features = ['pclass', 'sex', 'sibsp', 'parch', 'fare', 'embarked', 'who']
labels = 'survived'

## x, y 데이터 생성

In [57]:
x = df[features]
y = df[labels]

In [58]:
x.head()

pclass  sex  sibsp  parch     fare  embarked  who
0       3    1      1      0   7.2500         0    0
1       1    0      1      0  71.2833         1    1
2       3    0      0      0   7.9250         0    1
3       1    0      1      0  53.1000         0    1
4       3    1      0      0   8.0500         0    0

In [59]:
y.head()

0    0
1    1
2    1
3    1
4    0
Name: survived, dtype: int64

## 표준화 적용

In [70]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
x_scaled = scaler.fit_transform(x)
x_scaled[:5]

array([[ 0.82737724,  0.73769513,  0.43279337, -0.47367361, -0.50244517,
        -0.56883712, -0.74259613],
       [-1.56610693, -1.35557354,  0.43279337, -0.47367361,  0.78684529,
         1.00518113,  0.77148431],
       [ 0.82737724, -1.35557354, -0.4745452 , -0.47367361, -0.48885426,
        -0.56883712,  0.77148431],
       [-1.56610693, -1.35557354,  0.43279337, -0.47367361,  0.42073024,
        -0.56883712,  0.77148431],
       [ 0.82737724,  0.73769513, -0.4745452 , -0.47367361, -0.48633742,
        -0.56883712, -0.74259613]])

## 학습 / 검증 데이터셋 분할

In [60]:
from sklearn.model_selection import train_test_split

- test_size=0.2
- random_state=123
- stratify 옵션 지정

In [71]:
# 코드를 입력해 주세요
x_train, x_test, y_train, y_test = train_test_split(x_scaled, y, stratify=y, test_size=0.2, random_state=123)

In [72]:
# 검증코드
x_train.shape, x_test.shape

((712, 7), (179, 7))

<p><strong>[출력 결과]</strong></p><pre>((712, 6), (179, 6))</pre>

## KNN 알고리즘을 이용하여 예측

In [73]:
from sklearn.neighbors import KNeighborsClassifier

- 소수 3째자리까지 출력
- 최적의 K 값 찾기

In [74]:
# 코드를 입력해 주세요
for k in range(1, 11):
    knn = KNeighborsClassifier(n_neighbors=k)
    knn.fit(x_train, y_train)
    pred = knn.predict(x_test)
    acc = (pred == y_test).mean()
    print(f'k={k}, accuracy: {acc:.3f}')

k=1, accuracy: 0.771
k=2, accuracy: 0.777
k=3, accuracy: 0.799
k=4, accuracy: 0.816
k=5, accuracy: 0.804
k=6, accuracy: 0.793
k=7, accuracy: 0.793
k=8, accuracy: 0.816
k=9, accuracy: 0.821
k=10, accuracy: 0.832


## 정확도를 더 높일 수 있는 방법을 총 동원하여 80% 이상 정확도에 도전합니다!

### [tip]

- who 컬럼 추가
- StandardScaler() 활용